### PR TITLE
[Audio]: added streaming mode to SpectrogramToAudio

### PR DESCRIPTION
**Pull request by nasretdinovr**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR adds the streaming mode to SpectrogramToAudio class in this script `nemo/collections/audio/modules/transforms.py`.

**Collection**: audio

# Changelog

- Implemented streaming iSTFT in SpectrogramToAudio with a use_streaming toggle, buffering the previous frame and reusing istft; integrated into forward without changing offline behavior.
- Added a unit test that verifies streaming (center=False, rectangular window) matches offline reconstruction by processing frames one-by-one and finalizing the tail.

# Usage

Potential usage could be found it this test:

```
tests/collections/audio/test_audio_modules_transforms.py::TestAudioSpectrogram::test_streaming_istft_matches_offline_rectangular_center_false
```
# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [-] Did you add or update any necessary documentation?
- [-] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 1]
```diff
@@ -332,3 +332,75 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
+        hop_length = fft_length // hop_div
+
+        # Generate random signal
+        T = fft_length * 20 + 5  # non-multiple of hop to exercise edges
+        x = torch.randn(batch_size, num_channels, T)
+
+        # Rectangular window and STFT with center=False
+        rect_window = torch.ones(fft_length)
+        x_bc = x.reshape(batch_size * num_channels, T)
+        spec = torch.stft(
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py]: Why not use `AudioToSpectrogram` instead?

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 2]
```diff
@@ -332,3 +332,75 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
+        """
+        torch.manual_seed(123)
+        hop_length = fft_length // hop_div
+
+        # Generate random signal
+        T = fft_length * 20 + 5  # non-multiple of hop to exercise edges
+        x = torch.randn(batch_size, num_channels, T)
+
+        # Rectangular window and STFT with center=False
+        rect_window = torch.ones(fft_length)
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py]: Include at least one tapered window in the unit tests.

File: nemo/collections/audio/modules/transforms.py [hunk 3]
```diff
@@ -322,3 +344,100 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+        Returns:
+            Time-domain samples of shape (B, C, hop_length).
+        """
+        if not input.is_complex():
+            raise ValueError("Expected `input` to be complex dtype for streaming.")
+
+        # Lazily initialize buffers
+        self._init_stream_buffers(input)
+
+        B, C, F, Nf = input.size()
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: nit: `Nf` -> `num_frames`

File: nemo/collections/audio/modules/transforms.py [hunk 4]
```diff
@@ -322,3 +344,100 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+        assert F == self.num_subbands, f"Number of subbands F={F} not matching self.num_subbands={self.num_subbands}"
+        if Nf != 1:
+            raise ValueError("stream_update expects exactly one frame (N=1).")
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Log the actual value of `Nf` as well, will be easier to debug failures.

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 5]
```diff
@@ -332,3 +332,75 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
         # Input must be complex
         with pytest.raises(ValueError):
             s2a(input=torch.randn(1, 1, fft_length // 2 + 1, 100))
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize('fft_length', [256, 512])
+    @pytest.mark.parametrize('hop_div', [2, 4])
+    @pytest.mark.parametrize('batch_size', [1, 2])
+    @pytest.mark.parametrize('num_channels', [1, 2])
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py:340]: include non-default `magnitude_power` and `scale` parameters in tests as well

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 6]
```diff
@@ -332,3 +332,75 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
+        # Switch window to rectangular for streaming
+        s2a.use_streaming = True
+        s2a.window = rect_window
+        s2a.reset_streaming()
+
+        parts = []
+        for t in range(N):
+            # Feed one frame at a time: shape (B, C, F, 1)
+            frame = spec[..., t : t + 1]
+            out = s2a.stream_update(frame)
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py]: It seems using `forward` would be preferred here, since `stream_update` won't take care of `magnitude_power` and `scale`.

File: nemo/collections/audio/modules/transforms.py [hunk 7]
```diff
@@ -322,3 +344,100 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+
+        """
+        if not self._stream_initialized:
+            return torch.tensor((), device=self.window.device)
+
+        if self._prev_spec_frame is None:
+            return torch.tensor((), device=self.window.device)
+
+        # For center=False, emit final tail from last frame via istft
+        last_frame_spec = self._prev_spec_frame.unsqueeze(-1)
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Note:
Calling `stream_finalize` at the very end should be fine after using `forward` to process individual frames, since `self._prev_spec_frame` will be magnitude_power/scale transformed as expected in `forward` call before calling `stream_update`.

File: nemo/collections/audio/modules/transforms.py [hunk 8]
```diff
@@ -322,3 +344,100 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+    def stream_update(self, input: torch.Tensor) -> torch.Tensor:
+        """Consume exactly one spectrogram frame and return hop_length samples.
+
+        This method supports batched, multi-channel inputs and performs
+        overlap-add buffering internally. For each call, it will output
+        exactly `hop_length` time-domain samples (normalized by the
+        accumulated squared-window weights).
+
+        Args:
+            input: Complex spectrogram of shape (B, C, F, 1), where
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: It should be relatively easy to support chunks with num_frames>1 and output the corresponding multiple of hop length.
If not necessary, it's ok to skip for now.

File: nemo/collections/audio/modules/transforms.py [hunk 9]
```diff
@@ -322,3 +344,100 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+            # center=False: emit the first hop directly from a single-frame iSTFT
+            self._prev_spec_frame = current_frame
+            one_frame_spec = input  # (B, C, F, 1)
+            out = self.istft(one_frame_spec)
+            hop = self.hop_length
+            return out[..., :hop]
+
+        two_frame_spec = torch.stack([self._prev_spec_frame, current_frame], dim=-1)  # (B, C, F, 2)
+        # Use istft which internally handles window-sum-square normalization
+        out = self.istft(two_frame_spec)
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Let's chat offline about this bit.

File: nemo/collections/audio/modules/transforms.py [hunk 10]
```diff
@@ -212,6 +216,15 @@ def __init__(self, fft_length: int, hop_length: int, magnitude_power: float = 1.
         logging.debug('\tmagnitude_power: %s', magnitude_power)
         logging.debug('\tscale:           %s', scale)
 
+        # --- Streaming state (initialized lazily) ---
+        self._stream_initialized: bool = False
+        # Time-domain overlap-add buffers (initialized lazily)
+        self._ola_accum: Optional[torch.Tensor] = None
+        self._ola_weight: Optional[torch.Tensor] = None
+        # Kept for backward compatibility; not used in OLA implementation
+        self._prev_spec_frame: Optional[torch.Tensor] = None
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: This is not used anymore?

File: nemo/collections/audio/modules/transforms.py [hunk 11]
```diff
@@ -322,3 +346,115 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+            return
+        if shape_like.dim() != 4:
+            raise ValueError("Expected input of shape (B, C, F, N_frames) for streaming.")
+        B, C = shape_like.size(0), shape_like.size(1)
+        device = shape_like.device
+        # Real-valued buffers for accumulated time-domain samples and weights
+        dtype = torch.float32 if shape_like.dtype == torch.complex64 else torch.float64
+        self._ola_accum = torch.zeros(B, C, self.win_length, device=device, dtype=dtype)
+        self._ola_weight = torch.zeros(B, C, self.win_length, device=device, dtype=dtype)
+        self._prev_spec_frame = None
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Remove.

File: nemo/collections/audio/modules/transforms.py [hunk 12]
```diff
@@ -322,3 +346,115 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+        self._stream_initialized = True
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Is there need to track this state separately?
In current way, you need to take care of setting and resetting `self._stream_initialized`.

Can you define it using something like this:
```
@property
def _stream_initialized(self):
   ola_initialized = not (self._ola_accum is None or self._ola_weight is None)
   return _ola_initialized
```

(haven't tested, so may include typos)

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 13]
```diff
@@ -332,3 +332,67 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
+        """
+        torch.manual_seed(123)
+        hop_length = fft_length // hop_div
+
+        # Generate random signal
+        T = fft_length * 20 + 5  # non-multiple of hop to exercise edges
+        x = torch.randn(batch_size, num_channels, T)
+
+        # Window and STFT with center=False
+        window = torch.ones(fft_length)
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py]: Check an alternative tapered window.

File: nemo/collections/audio/modules/transforms.py [hunk 14]
```diff
@@ -322,3 +346,115 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+
+    @torch.no_grad()
+    def stream_finalize(self) -> torch.Tensor:
+        """Flush the remaining buffered samples (final tail for center=False).
+
+        After processing the last frame, the streaming loop has emitted N*hop
+        samples. The remaining tail corresponds to the last (win_length - hop)
+        samples, which we return after proper window-sum-square normalization.
+        """
+        if not self._stream_initialized or self._ola_accum is None or self._ola_weight is None:
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: Are all these three necessary?
Would it be sufficient to check `not self._stream_initialized`?

File: nemo/collections/audio/modules/transforms.py [hunk 15]
```diff
@@ -322,3 +346,115 @@ def get_output_length(self, input_length: torch.Tensor) -> torch.Tensor:
...
+            return torch.tensor((), device=self.window.device)
+
+        tail_len = self.win_length - self.hop_length
+        if tail_len <= 0:
+            return torch.tensor((), device=self.window.device)
+
+        eps = torch.finfo(self._ola_weight.dtype).eps
```
Comment (anteju) [inline nemo/collections/audio/modules/transforms.py]: This is repeated a couple times -- maybe make it a property, or a constant.

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 16]
```diff
@@ -332,3 +332,67 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
             s2a(input=torch.randn(1, 1, fft_length // 2 + 1, 100))
+
+    @pytest.mark.unit
+    @pytest.mark.parametrize('fft_length', [256, 512])
+    @pytest.mark.parametrize('hop_div', [2, 4])
+    @pytest.mark.parametrize('batch_size', [1, 2])
+    @pytest.mark.parametrize('num_channels', [1, 2])
+    @pytest.mark.parametrize('magnitude_power', [0.5, 1, 2])
+    @pytest.mark.parametrize('scale', [0.1, 1.0])
+    @pytest.mark.parametrize('chunk_size', [1, 2, 5])
```
Comment (anteju) [inline tests/collections/audio/test_audio_modules_transforms.py:343]: 👍

File: tests/collections/audio/test_audio_modules_transforms.py (hunk 1 above)
Comment (nasretdinovr) [reply tests/collections/audio/test_audio_modules_transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 3 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 4 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: no longer needed

File: tests/collections/audio/test_audio_modules_transforms.py (hunk 6 above)
Comment (nasretdinovr) [reply tests/collections/audio/test_audio_modules_transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 8 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 7 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: could you please clarify this?

File: nemo/collections/audio/modules/transforms.py (hunk 9 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 7 above)
Comment (anteju) [reply nemo/collections/audio/modules/transforms.py]: There's no ask here -- it was a clarifying note for self while reviewing.

File: nemo/collections/audio/modules/transforms.py (hunk 10 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: Correct, removed

File: nemo/collections/audio/modules/transforms.py (hunk 11 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: nemo/collections/audio/modules/transforms.py (hunk 12 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: tests/collections/audio/test_audio_modules_transforms.py (hunk 13 above)
Comment (nasretdinovr) [reply tests/collections/audio/test_audio_modules_transforms.py]: added hamming_window

File: nemo/collections/audio/modules/transforms.py (hunk 15 above)
Comment (nasretdinovr) [reply nemo/collections/audio/modules/transforms.py]: done

File: tests/collections/audio/test_audio_modules_transforms.py [hunk 17]
```diff
@@ -332,3 +332,89 @@ def test_invalid_spec_to_audio_input(self, fft_length: int):
...
+        spec2audio.reset_streaming()
+
+        parts = []
+        for t in range(0, N, chunk_size):
+            # Feed chunk_size frames at a time: shape (B, C, F, K)
+            frames = spec[..., t : min(t + chunk_size, N)]
+            out, _ = spec2audio(input=frames)
+            parts.append(out)
+        tail = spec2audio.stream_finalize()
+        x_stream = torch.cat(parts + [tail], dim=-1)
```
Comment (racoiaws) [inline tests/collections/audio/test_audio_modules_transforms.py:414]: We really need documentation on how to use this new functionality. It is not obvious from methods names and their docstrings. I had to go to look up tests

Especially why we have to do
```
spec2audio.window = window
spec2audio.reset_streaming()
```
and
```
tail = spec2audio.stream_finalize()
x_stream = torch.cat(parts + [tail], dim=-1)
```

Comment (nasretdinovr) [reply tests/collections/audio/test_audio_modules_transforms.py:414]: added an example of streaming istft usage to the docstring of SpectrogramToAudio class
